modules/queue/queue_channel.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package queue
+
+import "time"
+
+// ChannelQueue implements
+type ChannelQueue struct {
+	queue  chan *Data
+	handle HandlerFunc
+}
+
+// NewChannelQueue create a memory channel queue
+func NewChannelQueue(handle HandlerFunc, queueLength int) *ChannelQueue {
+	return &ChannelQueue{
+		queue:  make(chan *Data, queueLength),
+		handle: handle,
+	}
+}
+
+// Run starts to run the queue
+func (c *ChannelQueue) Run() error {
+	for {
+		select {
+		case data := <-c.queue:
+			c.handle(data)
+		}
+	}
+}
+
+// Push will push the indexer data to queue
+func (c *ChannelQueue) Push(data *Data) error {
+	c.queue <- data
+	return nil
+}
+
+// BatchedChannelQueue implements
+type BatchedChannelQueue struct {
+	*ChannelQueue
+	batchLength int
+}
+
+// NewBatchedChannelQueue create a memory channel queue
+func NewBatchedChannelQueue(handle HandlerFunc, batchLength, queueLength int) *BatchedChannelQueue {
+	return &BatchedChannelQueue{
+		&ChannelQueue{
+			queue:  make(chan *Data, queueLength),
+			handle: handle,
+		},
+		batchLength,
+	}
+}
+
+// Run starts to run the queue
+func (c *BatchedChannelQueue) Run() error {
+	var i = 0
+	var datas = make([]*Data, 0, c.batchLength)
+	for {
+		select {
+		case data := <-c.queue:
+			datas = append(datas, data)
+			if len(datas) >= c.batchLength {
+				c.handle(datas...)
+				datas = make([]*Data, 0, c.batchLength)
+			}
+		case <-time.After(time.Millisecond * 100):
+			i++
+			if i >= 3 && len(datas) > 0 {
```

Comment:
3 -> batchLength ?

Reply:
It's from your original code :stuck_out_tongue: 

My interpretation is:

* 300ms after you have started this queue, send any unsent data even if the batch is incompletely full and every 100ms after that.

Reply:
I've restructured to be a bit clearer - the timeout variable changes once it's had its first wait.